service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,140 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.dto.mapper.CertificateServiceMapper;
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.exception.EntityAlreadyExistsException;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.InvalidEntityException;
+import com.epam.esm.repository.SearchCriteria;
+import com.epam.esm.repository.repositoryinterfaces.CertificateRepository;
+import com.epam.esm.repository.repositoryinterfaces.TagRepository;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.validator.CertificateValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private final CertificateRepository certificateRepository;
+    private final TagRepository tagRepository;
+    private final CertificateServiceMapper certificateServiceMapper;
+    private final CertificateValidator certificateValidator;
+
+    public CertificateServiceImpl(CertificateRepository certificateRepository,
+                                  TagRepository tagRepository,
+                                  @Autowired CertificateServiceMapper certificateServiceMapper,
+                                  CertificateValidator certificateValidator) {
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+        this.certificateServiceMapper = certificateServiceMapper;
+        this.certificateValidator = certificateValidator;
+    }
+
+    @Override
+    public List<CertificateDto> findCertificateByCriteriaService(SearchCriteria searchCriteria) {
+
+        certificateValidator.isSearchCriteriaEmpty(searchCriteria);
+
+        List<Certificate> certificateList = certificateRepository.find(searchCriteria);
+
+
+        return certificateList.stream().map(certificate -> {
+            long id = certificate.getId();
+            List<Tag> byCertificateId = tagRepository.findByCertificateId(id);
+            return certificateServiceMapper.convertCertificateToDto(certificate, byCertificateId);
+        }).collect(Collectors.toList());
+    }
+
+    @Override
+    public CertificateDto findCertificateByIdService(long id) {
+        if (id < 1) {
+            throw new EntityNotFoundException(id);
+        }
+
+        Certificate certificate = certificateRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException(id));
+        List<Tag> byCertificateId = tagRepository.findByCertificateId(id);
+
+        return certificateServiceMapper.convertCertificateToDto(certificate, byCertificateId);
+    }
+
+    @Override
+    public boolean attachTagToCertificateService(long certificateId, long tagId) {
```

Comment:
```suggestion
    public boolean attachTagToCertificate(long certificateId, long tagId) {
```
apply for all methods

Reply:
fixed